libraries/Adapters/Microsoft.Bot.Builder.Adapters.Twilio/TwilioAdapter.cs

```diff
@@ -0,0 +1,243 @@
+﻿// Copyright(c) Microsoft Corporation.All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json;
+using Twilio;
+using Twilio.Rest.Api.V2010.Account;
+using Twilio.Security;
+
+namespace Microsoft.Bot.Builder.Adapters.Twilio
+{
+    public class TwilioAdapter : BotAdapter
+    {
+        private readonly ITwilioAdapterOptions _options;
+
+        public TwilioAdapter(ITwilioAdapterOptions options)
+        {
+            _options = options;
+
+            if (options.TwilioNumber.Equals(string.Empty))
+            {
+                throw new Exception("TwilioNumber is a required part of the configuration.");
+            }
+
+            if (options.AccountSid.Equals(string.Empty))
+            {
+                throw new Exception("AccountSid is a required part of the configuration.");
+            }
+
+            if (options.AuthToken.Equals(string.Empty))
+            {
+                throw new Exception("AuthToken is a required part of the configuration.");
+            }
+
+            TwilioClient.Init(_options.AccountSid, _options.AuthToken);
+        }
+
+        /// <summary>
+        /// Standard BotBuilder adapter method to send a message from the bot to the messaging API.
+        /// </summary>
+        /// <param name="turnContext">A TurnContext representing the current incoming message and environment.</param>
+        /// <param name="activities">An array of outgoing activities to be sent back to the messaging API.</param>
+        /// <param name="cancellationToken">A cancellation token for the task.</param>
+        /// <returns>A resource response.</returns>
+        public override async Task<ResourceResponse[]> SendActivitiesAsync(ITurnContext turnContext, Activity[] activities, CancellationToken cancellationToken)
+        {
+            var responses = new List<ResourceResponse>();
+            for (var i = 0; i < activities.Length; i++)
+            {
+                var activity = activities[i];
+                if (activity.Type == ActivityTypes.Message)
+                {
+                    var messageOptions = ActivityToTwilio(activity);
+
+                    var res = await MessageResource.CreateAsync(messageOptions);
+
+                    var response = new ResourceResponse()
+                    {
+                        Id = res.Sid,
+                    };
+
+                    responses.Add(response);
+                }
+                else
+                {
+                    // TODO: log error: unknown message type
```

Comment:
Resolve TODO, what should we do with no messages activities? Throw an exception or remove the else and continue?

Reply:
Done. Thanks!